tensorflow/core/kernels/data/batch_dataset_op_test_v2.cc

```diff
@@ -0,0 +1,760 @@
+        drop_remainder(CreateTensor<bool>(TensorShape({}), {drop_remainder})),
+        parallel_copy(parallel_copy) {}
+
+  Status MakeInputs(gtl::InlinedVector<TensorValue, 4>* inputs) override {
+    if (input_dataset.NumElements() == 0 ||
+        input_dataset.dtype() != DT_VARIANT) {
+      return tensorflow::errors::Internal(
+          "The input dataset is not populated as the dataset tensor yet.");
+    }
+    *inputs = {TensorValue(&input_dataset), TensorValue(&batch_size),
+               TensorValue(&drop_remainder)};
+    return Status::OK();
+  }
+
+  int64 num_input_elements;
+  Tensor input_dataset;
+  Tensor batch_size;
+  Tensor drop_remainder;
+  bool parallel_copy;
+};
+
+class BatchDatasetOpTest : public DatasetOpsTestBase {
+ public:
+  Status SetupTestEnv(int thread_num, int cpu_num,
+                      const std::vector<FunctionDef>& flib) {
+    TF_RETURN_IF_ERROR(InitThreadPool(thread_num));
+    TF_RETURN_IF_ERROR(InitFunctionLibraryRuntime(flib, cpu_num));
+  }
+
+  Status MakeDatasetAndIterator(BatchDatasetParams batch_dataset_params,
+                                bool create_iterator) {
+    batch_dataset_params_ = std::move(batch_dataset_params);
+
+    // Populate the `input_dataset` in `batch_dataset_params_`.
+    RangeDatasetParams range_dataset_params(
+        0, batch_dataset_params_.num_input_elements, 1, {DT_INT64},
+        {PartialTensorShape({})}, "");
+    TF_RETURN_IF_ERROR(MakeRangeDataset(
+        range_dataset_params.start, range_dataset_params.stop,
+        range_dataset_params.step, range_dataset_params.output_dtypes,
+        range_dataset_params.output_shapes,
+        &batch_dataset_params_.input_dataset));
+
+    // Create the dataset kernel.
+    TF_RETURN_IF_ERROR(CreateBatchDatasetOpKernel(batch_dataset_params_,
+                                                  &batch_dataset_kernel_));
+
+    // Create the inputs for the dataset op.
+    gtl::InlinedVector<TensorValue, 4> inputs;
+    TF_RETURN_IF_ERROR(batch_dataset_params_.MakeInputs(&inputs));
+    // Creat the dataset context.
+    TF_RETURN_IF_ERROR(CreateBatchDatasetContext(
+        batch_dataset_kernel_.get(), &inputs, &batch_dataset_context_));
+    // Create the dataset.
+    DatasetBase* batch_dataset;
+    TF_RETURN_IF_ERROR(CreateDataset(batch_dataset_kernel_.get(),
+                                     batch_dataset_context_.get(),
+                                     &batch_dataset));
+    batch_dataset_.reset(batch_dataset);
+
+    if (!create_iterator) {
+      return Status::OK();
+    }
+
+    // Create the iterator context.
+    TF_RETURN_IF_ERROR(
+        CreateIteratorContext(batch_dataset_context_.get(), &iterator_ctx_));
+    // Create the iterator.
+    TF_RETURN_IF_ERROR(batch_dataset_->MakeIterator(
+        iterator_ctx_.get(), kIteratorPrefix, &iterator_));
+  }
+
+  Status CheckIteratorGetNext(const std::vector<Tensor>& expected_outputs,
```

Comment:
The `Check*` functions will be moved to the test base once we finalize the solution.